ismatec/util.py

```diff
@@ -116,39 +109,28 @@ def join(self, timeout=None):
 class SerialCommunicator(Communicator):
     """Communicator using a directly-connected RS232 serial device."""
 
-    def init(self):
-        """Initialize serial port."""
-        assert type(self.address) == str
-        self.ser = serial.Serial(self.address, **self.serial_details)
+    def __init__(self, address, timeout=.05):
+        """Initialize the serial link and create queues for commands and responses."""
+        super(SerialCommunicator, self).__init__()
+        self.ser = serial.Serial(
+            address,
+            baudrate=9600,
+            bytesize=8,
+            parity='N',
+            stopbits=1,
+        )
+
+    def write(self, message : str):
+        """Write a message to the device."""
+        self.ser.write(message.encode() + b'\r')
+
+    def read(self, length : int):
+        """Read a fixed number of bytes from the device."""
+        return self.ser.read(length)
 
-    def loop(self):
-        """Do the repetitive work."""
-        # deal with commands and queries found in the request queue
-        if self.req_q.qsize():
-            # disable asynchronous communication
-            self.command(b'1xE0\r')
-            self.ser.read(1)
-            # empty the ingoing buffer
-            flush = self.ser.read(100)
-            if flush:
-                logger.debug(f'flushed garbage before query: "{flush!r}"')
-            # write command and get result
-            cmd = self.req_q.get()
-            self.command(cmd.encode() + b'\r')
-            res = self.ser.readline().strip()
-            self.res_q.put(res)
-            # enable asynchronous communication again
-            self.command(b'1xE1\r')
-            self.ser.read(1)
-        line = self.ser.readline()
-        if len(line):
-            # check for running message
-            if line[:2] == '^U':
-                ch = int(line[2])
-                self.running[ch] = True
-            elif line[:2] == '^X':
-                ch = int(line[2])
-                self.running[ch] = False
+    def readline(self):
+        """Read until a LF terminator."""
+        self.ser.readline().strip()
```

Comment:
Hahaha, I spent _way_ too long debugging the fact that this never returns the line it reads 0_0
@patrickfuller